### ISSUE TITLE
Write-Output -NoEnumerate doesn't handle Hashtable properly

### ISSUE THREAD

**Issue by kganjam** [PowerShell/PowerShell#14047]
## Steps to reproduce
```
$g = @{}
$g["foo"] = 1
$g["bar"] = 1
$g["foo"] # returns 1
$g."foo" # returns 1

$h = write-output $g -NoEnumerate
$h["foo"] # returns $null
$h."foo" # returns 1
$h["bar"] # returns $null
$h."bar" # returns 1
$h.count # returns 1
$($h.keys | measure).count # returns 2

$f = write-output $g
$f["foo"] # returns 1
$f."foo" # returns 1
$f["bar"] # returns 1
$f."bar" # returns 1
$f.count # returns 2
```

## Expected behavior

``$h["foo"] # returns 1``
``$h.count # returns 2``

## Actual behavior

``$h["foo"] # returns $null``
``$h.count # returns 1``

## Environment data
```
Name                           Value
----                           -----
PSVersion                      7.1.0
PSEdition                      Core
GitCommitId                    7.1.0
OS                             Microsoft Windows 10.0.19042
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

## Notes
I thought I would use `end { Write-Output $Result -NoEnumerate }` to be safe for `ConvertTo-Dictionary` the way I do to ensure an empty HashSet is returned by `ConvertTo-HashSet` when pipeline input is empty, but it turns out you must use `end { $Result }` or you get back a broken Hashtable. 

Maybe this is expected behavior somehow, but I don't understand what $h is and how $h."foo" and $h."bar" both work even though $h.count is 1. And why $h.keys gives 2 items, while $h.count is 1. I guess $h is being unrolled and that is how it is working. Still, not what I would have expected from -NoEnumerate. 
```
<#
.SYNOPSIS
Convert a stream of objects to a Dictionary
.EXAMPLE
$services = Get-Service | ConvertTo-Dictionary {$_.Name} {$_.DisplayName}
#>
function ConvertTo-Dictionary {
    param(
        [ScriptBlock]$KeyExpression,
        [ScriptBlock]$ValueExpression = { $_ }
    )
    begin { $Result = @{} }
    process { $Result[(&$KeyExpression)] = (&$ValueExpression) }
    end { $Result } 
}

<#
.SYNOPSIS
Convert a stream of objects to a HashSet
.EXAMPLE
$services = get-service | select -expandproperty Name -first 10 | ConvertTo-HashSet 
$null -eq $services # False
$Services = get-service | select -expandproperty Name -first 0 | ConvertTo-HashSet 
$null -eq $services # False
#>
function ConvertTo-HashSet {
    begin { $hashSet = [System.Collections.Generic.HashSet[String]] @() }
    process {
        $hashSet.add($_) | Out-Null
    }
    end { Write-Output -NoEnumerate $hashSet }
}
```

**Comment by kganjam** [comment]
**Update:** 
`write-output $g -NoEnumerate | gm`
returns
``
TypeName: System.Collections.Generic.List`1[[System.Object, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]
``
So I guess I now understand what NoEnumerate is doing. 

I guess my confustion stems from why @{} doesn't get converted to $null while an empty HashSet does. 

```
$(write-output $([System.Collections.Generic.HashSet[String]] @())) -eq $null # True
$(write-output $(@{})) -eq $null # False
```
They both seem to implement a similar set of interfaces.  

```
$(@{}).GetType().implementedInterfaces
$(write-output $([System.Collections.Generic.HashSet[String]] @()) -NoEnumerate).GetType().ImplementedInterfaces
```

I guess these threads help: 
https://stackoverflow.com/questions/28702588/in-what-conditions-does-powershell-unroll-items-in-the-pipeline
https://stackoverflow.com/questions/1827862/what-determines-whether-the-powershell-pipeline-will-unroll-a-collection

Perhaps this is a doc bug?

**Comment by vexx32** [comment]
A HashSet is more analogous to an array than a dictionary, it only contains values; a Hashtable is a dictionary, containing both keys and values.

Empty hashset -> empty list
Empty dictionary -> list with one item (the dictionary).

I'm not really sure whether a single item of input _should_ be wrapped in a List collection in this instance. My inclination is that it probably shouldn't, but it might be that the cmdlet has no good way of determining that necessarily... 🤔

**Comment by mklement0** [comment]
@kganjam:

Another way of putting it: even though `[hashtable]` or, more generally, `IDictionary`  implements / derives from `IEnumerable`, PowerShell has decided _not_ to enumerate it (its key-value pairs) in the pipeline.

The list of hard-coded exceptions (types that implement `IEnumerable` but aren't enumerated in the pipeline):
* types that implement `System.Collections.IDictionary`
* `System.Xml.XmlNode`
* `System.String` (which is technically an enumerable over its `System.Char` instances).
   
There's also an opposite exception: `System.Data.DataTable` _is_ enumerated, despite _not_ implementing `IEnumerable`.

See https://github.com/PowerShell/PowerShell/blob/f4ac0935c530be5b33c3fc043b574b95d41a1ca5/src/System.Management.Automation/engine/LanguagePrimitives.cs#L525


@vexx32:

> I'm not really sure whether a single item of input should be wrapped in a List collection in this instance. My inclination is that it probably shouldn't, but it might be that the cmdlet has no good way of determining that necessarily..

I'd say it's an obvious bug, one with a storied history: #5122 (the symptom has changed since then, but the fundamental problem is the same), possibly going all the way back to the ill-fated #2038

**Comment by vexx32** [comment]
Yeah, you're right there; #2038 was definitely where it originated, and pretty much ever since then it's been a matter of squashing the edge cases around it.

I'm not really sure of the best way to handle this kind of case specifically, though.

**Comment by mklement0** [comment]
This may be a moot point, but I've always believed #2038 to be conceptually flawed and that it should never have been implemented: https://github.com/PowerShell/PowerShell/issues/2035#issuecomment-323641345
It continues to cause trouble.

**Comment by vexx32** [comment]
I don't think I really disagree there. Not sure it's something we can undo now... but I'd agree that it's probably been a lot more trouble than it's worth.

**Comment by iSazonov** [comment]
If you think we need to revert this, please open new issue for discussion now before we start next LTS version.
